### PR TITLE
WIP: Immutable DTO for hocon

### PR DESCRIPTION
**Pull request by Arkatufus**
- Add immutable support library for materialized Hocon:
    - Class types are reduced to Object, Array, and Literal value type
    - Object value class emulates ImmutableDictionary, with a custom indexer that can directly take a Hocon path, with an ImmutableSortedDictionary as a backing data source
    - Array value class emulates ImmutableList, with an ImmutableArray as a backing data source.
    - Literal value is just a wrapper around a string to make it immutable.
- Improved data access API for the immutable tree:
    - Objects can be accessed directly using string or HoconPath path inside the dictionary-like indexer. Eg. `config["root.first-object.first-property"]` will treat config as an object value and traverse to the "root.first-object.first-property" path.
    - The library is smart enough to deduce data access from indexer type, integer indexer will be automatically interpreted as array access, string indexer will be automatically interpreted as path traversal. Eg. `config["root.my-array"][0]` will treat the `config` instance as a Hocon object, traverse it to the `"root.my-array"` value, and then the `[0]` indexer will treat that value as an array. It will read array values directly, or if it is an object value, will try to convert it into an array using the "positively numeric keyed object to array convertion" rule. Accessing a literal value using any indexers will throw an exception.
    - Literal values are automatically converted/parsed when accessed. Eg. `byte value = config["root.byte-value"]`, the value will be auto convert to byte and assigned to the variable. Auto convert will validate the conversion on the fly and throws an exception if the conversion is invalid.
    - Generic getter implementation. You can use the generic getter instead of relying on the auto type conversion. Eg. `var myIntArray = config["root.int-array"].Get<int[]>()` will return an array of ints from "root.int-array" path, or `var myTimeSpan = config["root.time-span"].Get<TimeSpan>()` will return a `TimeSpan` from "root.time-span" path.
    - Old style getters are still supported, you can still use `myConfig.Value.GetInt()`, `myConfig.Value.GetInt("my-path")` etc. ~~but they will only work on the Literal class instance, you can not pass a Hocon path into them anymore.~~ The `.Value` property is deprecated, it is included just to maintain backward compatibility with old code.
    - Auto-casting and generic getter only works for primitive types, there are no POCO class converter support yet.
    - Supported auto-cast and generic types: bool, byte, sbyte, short, ushort, int, uint, long, ulong, BigInteger, float, double, decimal, TimeSpan, string, and their array variant (eg. bool[] and byte[])

### CODE REVIEW

**Comment by Arkatufus** [comment]
I'm maintaining full backward compatibility with the old API. Should this be a separate library or should it be folded back into the core project? Ie, should config immutability be the default?

**Comment by iron9light** [comment]
You wanna re-implemented with all immutable data struct?
That's a good idea. It will be easier to do things right and easier to test.
I think backward compatibility is not a must-have.
You can make it HOCON 2.0 and keep maintain a HOCON 1.x branch for bug fixing.

**Comment by Arkatufus** [comment]
The original HOCON API are all immutable, I'm just trying to see if we should stick to the original material, plus immutability does have its advantages. Programmatic HOCON building/materialization is also supported using a builder pattern, instead of using the fluent .With pattern, which I kind of hate, because it is so memory intensive.

File: src/Hocon.Immutable/Extensions/GenericGetterExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hocon.Immutable.Extensions
+{
+    public static class GenericGetterExtensions
```
Comment (Arkatufus) [inline src/Hocon.Immutable/Extensions/GenericGetterExtensions.cs]: Generic value getter implementation. Converts an immutable Hocon literal string value to another primitive type. Usage: `myHoconLiteral.Get<int>()`

File: src/Hocon.Immutable/Extensions/ValueGetterExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,340 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Hocon.Immutable.Extensions
+{
+    public static class ValueGetterExtensions
```
Comment (Arkatufus) [inline src/Hocon.Immutable/Extensions/ValueGetterExtensions.cs]: Contains static extension convenience helper functions to convert an immutable hocon literal into another primitive type. Sample use: `myHoconLiteral.GetDecimal()`.

Also contains backward compatible getter helper functions that follows the old `myHoconObject.GetInt("my.hocon.path")` API pattern.

File: src/Hocon.Immutable.Tests/ApiTest.cs [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+    timespan: 1m
+    null : null
+  },
+
+  root_2 : 1234
+}";
+            var config = Parser.Parse(hocon).ToHoconImmutable();
+            Assert.True(config["root.bool"].GetBoolean());
+            Assert.True(config["root.bool"].Get<bool>());
+            Assert.True(config["root.bool"]);
```
Comment (Arkatufus) [inline src/Hocon.Immutable.Tests/ApiTest.cs]: Sample use of the auto-conversion getter. `ImmutableHoconLiteral` contains implicit cast operator overrides to automatically convert the string value into other primitive types. In this case, it automatically converts the string value to `bool`

**Comment by Aaronontheweb** [comment]
Have this on my to-do list this Friday

File: src/Hocon.Immutable.Tests/ApiTest.cs [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+{c=3,d=4}, 
+{e=5,f=6},]"},
+                new object[]{@"a=[
+{a=1,b=2}, 
+{c=3,d=4}, 
+{e=5,f=6}]"}
+            };
+        [Theory]
+        [MemberData(nameof(ObjectArrayData))]
+        public void CanIntelligentlyUseIndexerTypeToAccessMembers(string hocon)
```
Comment (Aaronontheweb) [inline src/Hocon.Immutable.Tests/ApiTest.cs]: Very cool - this is a slick addition to the API

File: src/Hocon.Immutable.Tests/ApiTest.cs [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+    string-array: [""1"",""2"",""3""]
+
+    timespan: 1m
+    null : null
+  },
+
+  root_2 : 1234
+}";
+            var config = Parser.Parse(hocon).ToHoconImmutable();
+            Assert.True(config["root.bool"].GetBoolean());
```
[the remark this answers is not in the document]
Comment (Aaronontheweb) [reply src/Hocon.Immutable.Tests/ApiTest.cs]: LGTM

**Comment by Arkatufus** [comment]
> Question though - why ship it as a separate project? Why not make it part of the whole?

That was what I asked earlier, should I fold this back into the main package or should I keep it separate into an opt-in package.

The main reason I split it up is to make a proof of concept that it works without having to touch the main code, to have flexibility, just in case it is better to make it an opt-in package; I was not planning on making these backward compatible, but it turns out that making it backward compatible was really trivial.

**Comment by Aaronontheweb** [comment]
Got it - let’s keep it in its own project for now and do a POC while integrating it into Akka.NET

**Pull request merged**